time_span/time_span.go

```diff
@@ -2,28 +2,210 @@ package time_span
+	if t.End.After(t.Start) {
+		return t.End.Sub(t.Start)
+	} else {
+		return time.Duration(0)
+	}
+}
+
+// Determines whether the interval includes the given query.
+func (t OpenInterval) Includes(query time.Time) bool {
+	return t.Start.Before(query) && query.Before(t.End)
+}
+
+// TODO: overlap between open and closed?
+
+// Determine whether this interval overlaps with another open interval
+// If one of them is empty, this should still return false.
+func (t1 OpenInterval) Overlaps(t2 OpenInterval) bool {
+	// Need C such that s1 < C < e1 and s2 < C < e2, so it is necessary
+	// that s1 < e2 and s2 < e1, but also that both of them are nonempty.
+	return t1.Start.Before(t2.End) && t2.Start.Before(t1.End) &&
+		t1.Start.Before(t1.End) && t2.Start.Before(t2.End)
+}
+
+// An half-open interval [start,end)
+type HalfOpenInterval BaseInterval
+
+func NewHalfOpenInterval(start time.Time, end time.Time) HalfOpenInterval {
+	return HalfOpenInterval{
+		Start: start,
+		End:   end,
+	}
+}
+
+func (span HalfOpenInterval) Duration() time.Duration {
+	if span.End.After(span.Start) {
+		return span.End.Sub(span.Start)
+	} else {
+		return time.Duration(0)
+	}
+}
+
+// Determines whether the interval includes the given query.
+func (span HalfOpenInterval) Includes(query time.Time) bool {
+	return (span.Start.Equal(query) || span.Start.Before(query)) && query.Before(span.End)
+}
+
+func (t1 HalfOpenInterval) Overlaps(t2 HalfOpenInterval) bool {
+	// There's some time C such that start1 <= C < end1 and start2 <= C < end2
+	// iff start1 < end2 AND start2 < end1.
+	return t1.Start.Before(t2.End) && t2.Start.Before(t1.End)
+}
+
+// Returns true if t2 is entirely contained within t1
+func (t1 HalfOpenInterval) Contains(t2 HalfOpenInterval) bool {
+	return (t1.Start.Before(t2.Start) || t1.Start.Equal(t2.Start)) &&
+		(t1.End.After(t2.End) || t1.End.Equal(t2.End))
+}
+
+// Extend an interval by "delta" in each direction (thus it gets 2*delta longer.)
+func (t HalfOpenInterval) Expand(delta time.Duration) HalfOpenInterval {
+	return HalfOpenInterval{
+		Start: t.Start.Add(time.Duration(-1) * delta),
+		End:   t.End.Add(delta),
+	}
+}
+
+// Return the intersection of two half-open intervals.
+// (i.e., trim this interval to lie entirely within another.)
+// May result in an interval with end <= start, which should be treated as empty.
+func (t HalfOpenInterval) Intersect(t2 HalfOpenInterval) HalfOpenInterval {
+	return HalfOpenInterval{
+		Start: MaxTime(t.Start, t2.Start), // later of the two starts
+		End:   MinTime(t.End, t2.End),     // earlier of the two ends
+	}
+}
+
+// Return the "union" of two half-open intervals, along with everything in between.
+// TOOD: I don't know if this has a nice mathematical name.
+func (t HalfOpenInterval) Combine(t2 HalfOpenInterval) HalfOpenInterval {
+	return HalfOpenInterval{
+		Start: MinTime(t.Start, t2.Start), // earlier of the two starts
+		End:   MaxTime(t.End, t2.End),     // later of the two ends
+	}
+}
```

Comment:
Combining two empty intervals can result in a non-empty interval:
```
empty1 := [1, 0)
empty2 := [3, 2)
empty1.Combine(empty2) == [1, 2)
```
Similarly, combining an empty interval with a non-empty interval shouldn't result in a different non-empty interval.
```
empty := [1, 0)
nonEmpty := [2, 3)
nonEmpty.Combine(empty) == [1, 3)
```

Reply:
Right, the problem is that this operation isn't a union, it's something weird.

The use case is: we have an event at [0,1) and we have an event at [4,5) and we want to merge it to a single interval.

If we have [0,0) instead then I would still want to merge to [0,5) even though the former is empty.

If we have [1,0) somehow then I am actually fine with [1,5) as the combination. I don't much care.

I would be happy to take this back out of the library so I can get the semantics I want for the combination, rather than arguing about what it would be in general.

Reply:
I'd advocate more for clear documentation than carefully figuring out what the semantics needs to be. For the cases where we are uncertain or don't currently care, we can just say that the result is undefined.